### PR TITLE
Network client side performance improvement. FrameData and OrderIO.

### PR DESCRIPTION
**Pull request by anvilvapre**
FrameData, cache IsReadyForFrame result, since called twice per net frame.
FrameData, avoid OrderBy client id.
FrameData, avoid LINQ.
FrameData, avoid Dictionary creation.
OrderIO, ToOrderList, skip if frame is empty, which occurs often each client each frame.

### CODE REVIEW

File: OpenRA.Game/Network/FrameData.cs [hunk 1]
```diff
@@ -56,26 +58,53 @@ public void AddFrameOrders(int clientId, int frame, byte[] orders)
...
+			}
+
+			if (ready)
+				frameData.Add(ClientIdAllReady, null);
+			return ready;
 		}
 
 		public IEnumerable<int> ClientsNotReadyForFrame(int frame)
-		{
+                {
```
Comment (deleted-user-1) [inline OpenRA.Game/Network/FrameData.cs]: style-nit: spaces but should use tabs

**Comment by deleted-user-1** [comment]
This was updated.

File: OpenRA.Game/Network/FrameData.cs (hunk 1 above)
Comment (anvilvapre) [reply OpenRA.Game/Network/FrameData.cs]: `make check` didn't notice, nor did i.

**Comment by adammitchelldev** [comment]
This will conflict with the changes proposed in #18783, which updates `FrameData.cs`.

**Comment by anvilvapre** [comment]
This issue introduces some smaller fixes. Perhaps it could be merged before the other that has bigger structural changes/improvements.

I assume the other issue addresses the last 2 items in commit message only. If they are all addresses we could close this one.

**Comment by anvilvapre** [comment]
See also #19019, unrelated to these changes, server side.

**Comment by adammitchelldev** [comment]
I more or less rewrote FrameData in my PR, so if we merged this first it
would basically have no effect, unless I were to rewrite my rewrite of
FrameData.

**Comment by reaperrr** [comment]
Maybe the `OrderIO` change could be split off to its own PR so it doesn't get held up by the discussion about how to proceed with the rest?

**Comment by reaperrr** [comment]
After discussions in IRC, we definitely want #18783.
To avoid rebase conflicts and redundant updates, I compared the `FrameData.cs` changes from both PRs, stopwatch-profiled the parts from here that were applicable there, with the following results:
- only the 'Avoid LINQ' changes from here looked applicable to #18783 
- the `IsReadyForFrame` changes showed only a miniscule improvement of 2-3 stopwatch ticks per OrderFrame tick, and at least when I tested showed a small perf regression on RA shellmap tick 0. Might have been an outlier, but overall the change didn't feel worth porting to #18783.
- the `OrdersForFrame` changes showed a notable improvement of 1/3 less CPU time (about 0.8 ms on my system) on RA shellmap tick 0 (potentially helping in perceived loading time), and a less consistent, but overall measurable reduction from about 50-60 to 30-40 stopwatch ticks on more regular order frames. I therefore suggested it in adapted form in #18783.

Due to the above, I think the `FrameData.cs` changes in this PR can be dropped, as the only worthwhile one applicable to #18783 will be applied there directly.

That would leave the change in `OrderIO`, and I measured that to actually have a more significant relative impact outside tick 0 (where it gives a tiny regression, insignificant compared to the speedup from `OrdersForFrame`, though) than the changes to `OrdersForFrame`, reducing stopwatch ticks by another ~60% on regular order frames on my system.
In absolute terms it's still a tiny gain, but combined with the `OrdersForFrame` changes it might still amount to something, at least on slower systems.

So 👍 once the `FrameData.cs` changes are removed.

**Comment by anvilvapre** [comment]
Performance aside, `OrdersForFrame` contained a bug. It doesn't make sense to create a list of all clients, sorts them to after find any entry. And to also call this methods twice each logic tick. 

All other changes are nice to have. The OrderIO one might be less relevant due to  #18783. I hope it will cause less often for empty order frames to be sent just to be able to sync.

The alternative fix to IsReadyForFrame was to maintain an extra boolean in the calling class. To avoid the double call.

**Comment by reaperrr** [comment]
> All other changes are nice to have. The OrderIO one might be less relevant due to #18783. I hope it will cause less often for empty order frames to be sent just to be able to sync.

I measured with that change applied to #18783, so it still has an effect there.

**Comment by anvilvapre** [comment]
Closes https://github.com/OpenRA/OpenRA/issues/13580

**Comment by pchote** [comment]
#19402 removed FrameData completely, so this will need to be rebased before we can proceed!

**Pull request closed without merging**